### PR TITLE
Flash Attention for Neuron

### PR DESCRIPTION
**Pull request by apoorvtintin**
This PR adds support for flash attention kernel for Neuron implemented through Neuron Kernel Interface (NKI).

The flash attention kernel works with TRN1 and TRN2.

This PR is a newer version of https://github.com/apple/axlearn/pull/883 from a different fork. All comments from the previous PR are addressed in this one. It has dropout support.

Dropout and Segment ID support in the flash attention kernel is in progress and will be available at a later date.

### CODE REVIEW

**Review by kelvin-zou — COMMENTED** [review]
Maybe wait until this PR is checked in. From what i can tell, your PR also has the remat bug not fixed. https://github.com/apple/axlearn/pull/942#pullrequestreview-2566173539

File: axlearn/common/flash_attention/utils.py [hunk 1]
```diff
@@ -276,13 +279,33 @@ def get_segment_ids(segment_ids: SegmentIdAttentionBias) -> Optional[Tensor]:
                 block_size=block_size,
             )
 
+        elif backend == "neuron":
```
Comment (ruomingp) [inline axlearn/common/flash_attention/utils.py:282]: Maybe only import `neuron_attention` inside this branch?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    causal: bool = False,
+    softmax_scale: float = 1.0,
+    dropout_rate: float = 0.0,
+):
+    out, _ = _mha_forward(query, key, value, bias, causal, softmax_scale, dropout_rate)
+    return out
+
+
+def _mha_forward(query, key, value, bias, causal, softmax_scale, dropout_rate):
+    # Get the batch size, sequence lengths, number of heads, and hidden dimension
```
Comment (ruomingp) [inline axlearn/common/flash_attention/neuron_attention.py:34]: Nit: end comments with . (here and everywhere)

File: axlearn/common/flash_attention/neuron_attention.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+@partial(custom_vjp, nondiff_argnums=(4, 5, 6))
+def flash_attention(
+    query: Tensor,
+    key: Tensor,
+    value: Tensor,
+    bias: Tensor,
+    causal: bool = False,
+    softmax_scale: float = 1.0,
+    dropout_rate: float = 0.0,
+):
```
Comment (ruomingp) [inline axlearn/common/flash_attention/neuron_attention.py:28]: Add comments on args and return value shapes and semantics?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
+# Copyright © 2024 Amazon Inc.
+"""Flash attention Kernels using NKI on Neuron. Tested on trn1 & trn2."""
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+
+# Import needed to enable JAX cache on Neuron
+import jax_neuronx  # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:9]: These packages are not publicly released yet, We will create another PR to update pyproject toml once they are. I will make them on demand import for the time being.

File: axlearn/common/flash_attention/neuron_attention.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    # Transpose the query, key, and value tensors
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len]
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len]
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model]
+
+    seed = jnp.array([1])
+
+    # Call the NKI kernel, duplicate the kernel if we cannot shard on num_heads
+    if (num_heads % 2) == 0 and (num_heads // 2 > 0):
```
[the remark this answers is not in the document]
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:45]: The tests in `neuron_attention.py` can run with a single attention head as well. This branch covers this corner case.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kelvin-zou) [reply axlearn/common/flash_attention/neuron_attention.py:9]: Can you then make it an on-demand import instead?

[the remark this answers is not in the document]
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:9]: Made neuron attention an on demand import everywhere.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 3 above)
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:28]: Added comments, let me know if they are sufficient.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:72]: Yeah this branch is needed for test support

File: axlearn/common/flash_attention/neuron_attention.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+lnc = 2 if jax.devices()[0].device_kind == "NC_v3d" else 1
+
+
+@partial(custom_vjp, nondiff_argnums=(4, 5, 6))
+def flash_attention(
+    query: Tensor,
+    key: Tensor,
+    value: Tensor,
+    bias: Tensor,
+    causal: bool = False,
```
Comment (kelvin-zou) [inline axlearn/common/flash_attention/neuron_attention.py:25]: Can we support segment ID? Or a more general masking fn (with optimized handling) is even better.

Comment (kelvin-zou) [reply axlearn/common/flash_attention/neuron_attention.py:25]: If not, I am fine with leaving a TODO here, but it is a hard blocker for enabling it for our internal training.

**Comment by apoorvtintin** [comment]
Thanks for all the reviews @ruomingp @kelvin-zou. I resolved all the comments, please let me know if any more changes are needed.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 6 above)
Comment (indhub) [reply axlearn/common/flash_attention/neuron_attention.py:25]: Can we do segment IDs in a separate PR? That involves non-trivial work and needs some time.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 3 above)
Comment (ruomingp) [reply axlearn/common/flash_attention/neuron_attention.py:28]: Thanks!

**Review by ruomingp — APPROVED** [review]

File: axlearn/common/flash_attention/neuron_attention.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+    # Transpose the query, key, and value tensors.
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len].
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len].
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model].
+
+    seed = jnp.array([1])
+
+    # Call the NKI kernel, duplicate the kernel if we cannot shard on num_heads.
+    if (num_heads % 2) == 0 and (num_heads // 2 > 0):
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:72]:
```
// even num_heads except 0
if num_heads > 0 and num_heads % 2 == 0:
```

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+def test_fwd_against_ref(
+    batch_size: int,
+    seq_len: int,
+    num_heads: int,
+    per_head_dim: int,
+    causal: bool,
+    input_dtype: jnp.dtype,
+    attention_bias_type: bool,
+):
+    softmax_scale = 1.0 / (per_head_dim**0.5)
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention_test.py:43]: Maybe just
```
per_head_dim**-0.5
```

**Comment by apoorvtintin** [comment]
I rebased the PR to avoid merge conflicts, can I please get a new approval? Thank you!

**Review by kelvin-zou — APPROVED** [review]

File: axlearn/common/flash_attention/neuron_attention.py (hunk 6 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/neuron_attention.py:25]: Sure, in this regard, I may ask for more, let's do general mask then, since we have want things beyond causal.

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+import chex
+import jax
+import jax.numpy as jnp
+import pytest
+
+from axlearn.common.flash_attention.utils import mha_reference
+
+if jax.default_backend() != "neuron":
+    pytestmark = pytest.mark.skip(reason="Incompatible hardware, AWS Neuron only test.")
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention_test.py:12]: Looks like a number of CI steps are failing -- I think we can either do something like 
```
if jax.default_backend() != "neuron":
    pytest.skip(reason=..., allow_module_level=True)
```
or update `run_tests.sh` to exclude tests marked with `neuron`.

**Comment by apoorvtintin** [comment]
I fixed the CI failures can we test it again? Thank you everyone!

**Comment by Ruixuan** [comment]
can we in the file, disable the pytype check with annotation like `# pytype: disable=import-error`
 

FAILED: /home/runner/work/axlearn/axlearn/.pytype/pyi/axlearn/common/flash_attention/neuron_attention.pyi 
/opt/hostedtoolcache/Python/3.10.16/x64/bin/python -m pytype.single --imports_info /home/runner/work/axlearn/axlearn/.pytype/imports/axlearn.common.flash_attention.neuron_attention.imports --module-name axlearn.common.flash_attention.neuron_attention --platform linux -V 3.10 -o /home/runner/work/axlearn/axlearn/.pytype/pyi/axlearn/common/flash_attention/neuron_attention.pyi --analyze-annotated --nofail --quick /home/runner/work/axlearn/axlearn/axlearn/common/flash_attention/neuron_attention.py
File "/home/runner/work/axlearn/axlearn/axlearn/common/flash_attention/neuron_attention.py", line 9, in <module>: Can't find module 'jax_neuronx'. [import-error]
File "/home/runner/work/axlearn/axlearn/axlearn/common/flash_attention/neuron_attention.py", line 10, in <module>: Can't find module 'neuronxcc'. [import-error]
File "/home/runner/work/axlearn/axlearn/axlearn/common/flash_attention/neuron_attention.py", line 12, in <module>: Can't find module 'neuronxcc.nki.kernels.attention'. [import-error]

For more details, see https://google.github.io/pytype/errors.html#import-error
[10/540] check axlearn.cloud.common.types
[11/540] infer axlearn.cloud.common.config
[12/540] check axlearn.common.utils
ninja: build stopped: subcommand failed.
Computing dependencies
Analyzing 535 sources with 0 local dependencies
Leaving directory '.pytype'
Error: Process completed with exit code 1.

**Comment by apoorvtintin** [comment]
@Ruixuan I updated the PR with `# pytype: disable=import-error` and verified that pytype passes. Can I get a new approval and re-trigger the CI? Thank you!

File: axlearn/common/flash_attention/neuron_attention.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    # Transpose the input tensors.
+    o = o.transpose(0, 2, 3, 1)
+    dy = d_attn_output.transpose(0, 2, 3, 1)
+
+    # Transpose v tensor.
+    v = jnp.transpose(v, axes=(0, 1, 3, 2))
+    seed = jnp.array([1])
+
+    # Call the NKI kernel, duplicate the kernel if we cannot shard on num_heads.
+    if (num_heads % 2) == 0 and (num_heads // 2 > 0):
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:129]: This can be simplified to
```
# even num_heads except 0
if num_heads > 0 and num_heads % 2 == 0:
```

File: axlearn/common/flash_attention/neuron_attention.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+    # Transpose the query, key, and value tensors.
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len].
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len].
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model].
+
+    seed = jnp.array([1])
+
+    # Call the NKI kernel, duplicate the kernel if we cannot shard on num_heads.
+    if (num_heads % 2) == 0 and (num_heads // 2 > 0):
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:75]: This can be simplified to
```
# even num_heads except 0
if num_heads > 0 and num_heads % 2 == 0:
```

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    per_head_dim: int,
+    causal: bool,
+    input_dtype: jnp.dtype,
+    attention_bias_type: bool,
+):
+    # On demand import only if test is needed.
+    # pylint: disable=import-outside-toplevel
+    from axlearn.common.flash_attention.neuron_attention import flash_attention
+
+    softmax_scale = 1.0 / (per_head_dim**0.5)
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention_test.py:44]: Maybe just

per_head_dim**-0.5

File: axlearn/common/flash_attention/neuron_attention.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    """
+    # Get the batch size, sequence lengths, number of heads, and hidden dimension.
+    batch_size, _, num_heads, _ = query.shape
+
+    # Transpose the query, key, and value tensors.
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len].
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len].
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model].
+
+    seed = jnp.array([1])
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:72]: why seed is hardcoded to [1]?

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+@pytest.mark.parametrize("attention_bias_type", [None, "2d"])
+@pytest.mark.parametrize("input_dtype", [jnp.float16, jnp.bfloat16, jnp.float32])
+def test_fwd_against_ref(
+    batch_size: int,
+    seq_len: int,
+    num_heads: int,
+    per_head_dim: int,
+    causal: bool,
+    input_dtype: jnp.dtype,
+    attention_bias_type: bool,
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention_test.py:38]: attention_bias_type is str, not bool
```
attention_bias_type: str | None,
```

File: axlearn/common/flash_attention/utils.py [hunk 15]
```diff
@@ -275,6 +275,28 @@ def get_segment_ids(segment_ids: SegmentIdAttentionBias) -> Optional[Tensor]:
...
+            # pylint: disable=import-outside-toplevel
+            from axlearn.common.flash_attention.neuron_attention import (
+                flash_attention as neuron_flash_attention,
+            )
+
+            key = _repeat_kv_heads(query.shape[2], key)
+            value = _repeat_kv_heads(query.shape[2], value)
+
+            # other_biases inlcudes SegmentIdAttentionBias among other biases.
```
Comment (apivovarov) [inline axlearn/common/flash_attention/utils.py:287]: typo: inlcudes -> includes

File: axlearn/common/flash_attention/neuron_attention.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+):
+    """Wraps _mha_forward for custom vjp.
+
+    Args:
+        query: Query of shape [batch_size, target_length, num_heads, per_head_dim].
+        key: Key of shape [batch_size, source_length, num_heads, per_head_dim].
+        value: Value of shape [batch_size, source_length, num_heads, per_head_dim].
+        bias: Optional logit biases of shape [1, 1, target_length, source_length].
+        softmax_scale: Optional scale to apply to softmax. Defaults to 1.
+        causal: Whether to apply causal mask.
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:40]: Align docstring parameter order with actual function signature

File: axlearn/common/flash_attention/neuron_attention.py (hunk 10 above)
Comment (apivovarov) [reply axlearn/common/flash_attention/neuron_attention.py:129]: BTW, should we use actual lnc value here instead of hardcoded 2?

File: axlearn/common/flash_attention/neuron_attention_test.py (hunk 12 above)
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention_test.py:44]: I think `softmax_scale = 1.0 / (per_head_dim**0.5)` is more readable.

File: axlearn/common/flash_attention/neuron_attention.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+            use_causal_mask=causal,
+            mixed_precision=True,
+            dropout_p=dropout_rate,
+            softmax_scale=softmax_scale,
+        )
+
+    # Transpose the gradients back to the original shape.
+    d_query = d_query.transpose(0, 3, 1, 2)
+    d_key = d_key.transpose(0, 3, 1, 2)
+    d_value = d_value.transpose(0, 3, 1, 2)
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:174]: can you add comment with expected dims for each transpose - similar to what you did in _mha_forward?

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+            dropout_rate=0.0,
+        ).sum()
+
+    def ref_fn(q, k, v, bias, segment_ids):
+        return mha_reference(
+            q,
+            k,
+            v,
+            bias,
+            segment_ids,
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention_test.py:135]: `segment_ids` is set to None which is default param value for it. It can be removed from `ref_fn` signature for consistency with fn/flash_attention above

File: axlearn/common/flash_attention/neuron_attention_test.py [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+@pytest.mark.parametrize("input_dtype", [jnp.bfloat16, jnp.float16, jnp.float32])
+@pytest.mark.parametrize("attention_bias_type", [None, "2d"])
+def test_bwd_against_ref(
+    batch_size: int,
+    num_heads: int,
+    seq_len: int,
+    per_head_dim: int,
+    causal: bool,
+    input_dtype: jnp.dtype,
+    attention_bias_type: bool,
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention_test.py:100]: attention_bias_type is str, not bool

File: axlearn/common/flash_attention/neuron_attention_test.py (hunk 12 above)
Comment (apivovarov) [reply axlearn/common/flash_attention/neuron_attention_test.py:44]: I'd use -0.5 for consistency with existing codebase - gpu_attention_test.py  uses ` q.shape[-1] ** -0.5`
```
softmax_scale = q.shape[-1] ** -0.5
```
Also, format the code

**Comment by apoorvtintin** [comment]
Resolved all comments and fixed CI failures, @Ruixuan @kelvin-zou @ruomingp can we re-trigger the CI and merge this?

**Review by kelvinzou — APPROVED** [review]

**Comment by apoorvtintin** [comment]
Hey everyone, looks like this PR also needs a new approval for the merge.

File: axlearn/common/flash_attention/neuron_attention.py [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+        dropout_rate: Dropout rate. Default to 0.0 (no dropout).
+
+    Returns:
+        The attention outputs of shape [batch_size, target_length, num_heads, per_head_dim].
+    """
+    out, _ = _mha_forward(query, key, value, bias, prng_key, causal, softmax_scale, dropout_rate)
+    return out
+
+
+def _mha_forward(query, key, value, bias, prng_key, causal, softmax_scale, dropout_rate):
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention.py:53]: Missing types?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+
+    See also `_mha_backward` for the backward pass.
+
+    Args:
+        query: Input query.
+        key: Input key.
+        value: Input value.
+        bias: Input bias.
+        prng_key: PRNG key used for dropout. Must be specified when dropout_rate > 0.0.
+        causal: Input segment_ids.
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention.py:64]: Looks like a stale docstring?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+        assert dropout_rate < 1
+        assert prng_key is not None
+    else:
+        # Dummy unused key.
+        prng_key = jax.random.key(0)
+
+    # TODO(apoorvtintin): Pass rbg key to kernel directly when kernel is ready to accept it.
+    # Currenlty NKI kernel supports a single 32 bit key, temporarily override this till support
+    # for 128 bit keys is added. Till then dropout is not supported.
+    prng_key = jnp.array([1])
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention.py:86]: Any reason not to raise if user specifies a prng_key?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+    batch_size, _, num_heads, _ = query.shape
+
+    # Transpose the query, key, and value tensors.
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len].
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len].
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model].
+
+    if dropout_rate > 0:
+        assert dropout_rate < 1
+        assert prng_key is not None
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention.py:78]: Prefer to raise ValueError over assert. Assertion errors are for catching logic bugs.

File: axlearn/common/flash_attention/neuron_attention.py [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+        grid = batch_size, num_heads
+
+    if bias is not None:
+        assert (
+            bias.ndim == 4
+        ), f"Neuron flash_attention is only expecting bias.ndim = 4 but got {bias.ndim}"
+        assert bias.shape[0] == 1 and bias.shape[1] == 1, (
+            f"Bias is only supported when batch and num_heads are both 1, "
+            f"batch is {bias.shape[0]} and num_heads is {bias.shape[1]}"
+        )
```
Comment (markblee) [inline axlearn/common/flash_attention/neuron_attention.py:101]: Likewise for these.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 6 above)
Comment (markblee) [reply axlearn/common/flash_attention/neuron_attention.py:30]: Consider leaving a `TODO(apoorvtintin)`?

Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:30]: Segment IDs are supported in the kernel through the bias argument, bias takes in the segment ID mask as well.

Comment (markblee) [reply axlearn/common/flash_attention/neuron_attention.py:30]: I think the ask was to support segment ids without materializing the whole `[target_len,source_len]` bias.

**Review by markblee — APPROVED** [review]
Thanks for the patience, @apoorvtintin !

**Comment by apoorvtintin** [comment]
Fixed CI failure

File: axlearn/common/flash_attention/neuron_attention.py (hunk 6 above)
Comment (indhub) [reply axlearn/common/flash_attention/neuron_attention.py:30]: We are improving the kernel to support segment ids without materializing the bias which I expect to be available sometime this month. Can we use the bias argument to do sequence packing until then? We tested it on 4K sequences and saw no noticeable difference in perf.

@apoorvtintin : Can you add a TODO for this?

Comment (markblee) [reply axlearn/common/flash_attention/neuron_attention.py:30]: Sounds great. As long as it's being tracked on your side happy to get the PR merged as-is.

**Comment by apoorvtintin** [comment]
Added a TODO: for segment ID support in the Kernel.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 6 above)
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:31]: Added a TODO: for this.

**Comment by apoorvtintin** [comment]
Rebased the PR to latest axlearn/main

**Comment by kelvin-zou** [comment]
Verified in internal bump up PRs.

**Review by markblee — APPROVED** [review]
